cmd/snap/cmd_run.go

```diff
@@ -482,26 +483,70 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func getInfoAndApp(snapName, appName string, rev snap.Revision) (*snap.Info, *snap.AppInfo, error) {
+	info, err := getSnapInfo(snapName, rev)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	app := info.Apps[appName]
+	if app == nil {
+		return nil, nil, fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	}
+
+	return info, app, nil
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
+
+	snapRev := snap.R(0)
+	hint, inhibitInfo, err := isLocked(snapName)
 	if err != nil {
 		return err
 	}
+	inhibited := hint != runinhibit.HintNotInhibited
+	if inhibited {
+		snapRev = inhibitInfo.Previous
+	}
 
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	info, app, err := getInfoAndApp(snapName, appName, snapRev)
+	if err != nil {
+		return err
 	}
 
 	if !app.IsService() {
 		if err := maybeWaitWhileInhibited(snapName); err != nil {
 			return err
 		}
+
+		if inhibited {
```

Comment:
Then again, I'm not sure if this check is particularly useful, as it may have become inhibited after the `inhibited` variable was initially set, in which case I believe we'd want to run this block of code anyway, even if `inhibited == false`. Please correct me if I'm wrong though, and we only care about whether the snap was inhibited when this was initially checked.

Reply:
This a great point, Thank you! Actually after looking at this again I agree that the info updating should always happen as you explained, even if the app is not a `Service` because we don't want the new inhibition logic to affect services code path which should remain unaffected (and actually fail to start when the snap is inhibited).

@pedronis I would love to know you thoughts on this.

Reply:
updated!